src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.ProjectState.cs

```diff
@@ -292,87 +286,78 @@ public bool OnProjectRemoved(ProjectId id)
             {
                 // loading data can be cancelled any time.
                 var project = document.Project;
 
                 var version = await GetDiagnosticVersionAsync(project, cancellationToken).ConfigureAwait(false);
-                var serializer = new DiagnosticDataSerializer(_owner.AnalyzerVersion, version);
+                var serializerVersion = version;
                 var builder = new Builder(project, version);
 
-                if (!await TryDeserializeDocumentDiagnosticsAsync(persistentService, serializer, document, builder, cancellationToken).ConfigureAwait(false))
+                if (!TryGetDiagnosticsFromInMemoryStorage(serializerVersion, document, builder))
                 {
                     return DiagnosticAnalysisResult.CreateEmpty(project.Id, VersionStamp.Default);
                 }
 
                 return builder.ToResult();
             }
 
-            private async Task<DiagnosticAnalysisResult> LoadInitialProjectAnalysisDataAsync(IPersistentStorageService persistentService, Project project, CancellationToken cancellationToken)
+            private async Task<DiagnosticAnalysisResult> LoadInitialProjectAnalysisDataAsync(Project project, CancellationToken cancellationToken)
             {
                 // loading data can be cancelled any time.
                 var version = await GetDiagnosticVersionAsync(project, cancellationToken).ConfigureAwait(false);
-                var serializer = new DiagnosticDataSerializer(_owner.AnalyzerVersion, version);
+                var serializerVersion = version;
                 var builder = new Builder(project, version);
 
-                if (!await TryDeserializeProjectDiagnosticsAsync(persistentService, serializer, project, builder, cancellationToken).ConfigureAwait(false))
+                if (!TryGetProjectDiagnosticsFromInMemoryStorage(serializerVersion, project, builder))
                 {
                     return DiagnosticAnalysisResult.CreateEmpty(project.Id, VersionStamp.Default);
                 }
 
                 return builder.ToResult();
             }
 
-            private async Task SerializeAsync(IPersistentStorageService? persistentService, DiagnosticDataSerializer serializer, Project project, TextDocument? document, object key, string stateKey, ImmutableArray<DiagnosticData> diagnostics)
+            private void AddToInMemoryStorage(VersionStamp serializerVersion, Project project, TextDocument? document, object key, string stateKey, ImmutableArray<DiagnosticData> diagnostics)
             {
                 Contract.ThrowIfFalse(document == null || document.Project == project);
 
-                // try to serialize it
-                if (persistentService != null &&
-                    await serializer.SerializeAsync(persistentService, project, document, stateKey, diagnostics, CancellationToken.None).ConfigureAwait(false))
```

Comment:
I believe the persistent storage cache always existed to reduce the memory pressure, especially when there are large number of diagnostics for things such as broken references. Tagging @heejaechang in case he can provide more context.

Reply:
resolving as we have reasonable belief that this is likely ok now.